### PR TITLE
use canonical path for auth

### LINKED ISSUE #3214: etcdctl fails to get key sometimes(but curl can) when auth enabled
I have a user `u1`, which is granted the permission to read '/config/globalsetting/7/*':

```
$ ./bin/etcdctl -u u1:u1 get /config/globalsetting/7/
Error:  0: Insufficient credentials () [0]
$ curl http://u1:u1@127.0.0.1:4001/v2/keys/config/globalsetting/7/
{"errorCode":100,"message":"Key not found","cause":"/config","index":41}
```

### PR DESCRIPTION
**Pull request by yichengq**
From https://github.com/coreos/etcd/issues/3214#issuecomment-127432079

use canonical path to ensure auth path match works.

fixes #3214

### CODE REVIEW

**Comment by yichengq** [comment]
test has fixed.

File: pkg/pathutil/path.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// limitations under the License.
+
+package pathutil
+
+import "path"
+
+// CanonicalURLPath returns the canonical url path for p, which follows the rules:
+// 1. the path always starts with "/"
+// 2. replace multiple slashes with a single slash
+// 3. replace each '.' '..' path name element with equivalent one
```
Comment (xiang90) [inline pkg/pathutil/path.go]: hm... do we need 3 for http request?

Comment (yichengq) [inline pkg/pathutil/path.go]: Yes.

HTTP request parses '.' as current directory and '..' as parent directory. The sentence from http.ServeMux description: `ServeMux also takes care of sanitizing the URL request path, redirecting any request containing . or .. elements to an equivalent .- and ..-free URL.` also confirms this. To get the canonical path for the same target, it needs to parse '.' and '..'

File: client/keys.go [hunk 2]
```diff
@@ -445,7 +445,13 @@ func (hw *httpWatcher) Next(ctx context.Context) (*Response, error) {
...
 // (typically "/v2/keys").
 func v2KeysURL(ep url.URL, prefix, key string) *url.URL {
-	ep.Path = path.Join(ep.Path, prefix, key)
+	if prefix != "" && prefix[0] != '/' {
+		prefix = "/" + prefix
+	}
+	if key != "" && key[0] != '/' {
+		key = "/" + key
+	}
+	ep.Path = pathutil.CanonicalURLPath(ep.Path + prefix + key)
```
Comment (xiang90) [inline client/keys.go]: I failed to see the different between pathutil.CanonicalURLPath(path.Join(ep.Path, prefix, key)).

I think this worth a comment + test.

Comment (yichengq) [inline client/keys.go]: comment and test added. The original code already has some related tests, e.g., empty prefix and key, and I add missed ones.

File: client/keys.go [hunk 3]
```diff
@@ -445,7 +445,16 @@ func (hw *httpWatcher) Next(ctx context.Context) (*Response, error) {
 // provided endpoint's path to the root of the keys API
 // (typically "/v2/keys").
 func v2KeysURL(ep url.URL, prefix, key string) *url.URL {
-	ep.Path = path.Join(ep.Path, prefix, key)
+	// This ensures slash at the head if prefix/key is not empty.
```
Comment (xiang90) [inline client/keys.go]: the head is actually in ep.Path. The difference I see here is to ensure we do not remove the trailing slash which path.Join would.

Comment (yichengq) [inline client/keys.go]: i should write out the reason why we don't use path.Join. will rewrite this.

**Comment by yichengq** [comment]
Could you read the comment again to see if it is clear enough?

File: client/keys.go [hunk 4]
```diff
@@ -445,7 +445,15 @@ func (hw *httpWatcher) Next(ctx context.Context) (*Response, error) {
 // provided endpoint's path to the root of the keys API
 // (typically "/v2/keys").
 func v2KeysURL(ep url.URL, prefix, key string) *url.URL {
-	ep.Path = path.Join(ep.Path, prefix, key)
+	// It concatenates all parts together manually. It cannot use
```
Comment (xiang90) [inline client/keys.go]: We .... We cannot use... it does not preserve the trailing slash. We call CanonicalURL to further cleanup the path.

File: pkg/pathutil/path.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// 4. keep the trailing slash
+func CanonicalURLPath(p string) string {
```
Comment (xiang90) [inline pkg/pathutil/path.go]: I think this is a directly copy from golang. We should use the their copyright i guess. /cc @philips

Comment (philips) [inline pkg/pathutil/path.go]: It is one of those functions that borderlines on obvious. I don't know if it is worth the hassle.

Comment (yichengq) [inline pkg/pathutil/path.go]: @xiang90 Another option is to leave a line to say `xxx borrowed from stdlib xxx` as we do it here https://github.com/coreos/etcd/blob/master/proxy/reverse.go#L35.

File: client/keys.go (hunk 4 above)
Comment (yichengq) [inline client/keys.go]: changed

**Comment by yichengq** [comment]
@ecnahc515 gives the suggestion to think about `ep.Parse/ResolveReference` function, but I think it does not work well in this case.
1. The relative resolve algorithm, which supports ResolveReference, is an algorithm for converting a URI reference that might be relative to a given base URI into the parsed components of the reference's target[1]. It doesn't guarantee a normalized path.
2. it doesn't parse the exactly the same way as http.ServeMux does, and ServeMux provides the request path that could be got by ServeHTTP. The different is that RelativeResolveAlgorithm keeps '//', while ServeMux replace it with single slash.

So i tend to keep the original implementation.

[1]: https://www.ietf.org/rfc/rfc3986.txt

File: pkg/pathutil/path.go (hunk 5 above)
Comment (yichengq) [inline pkg/pathutil/path.go]: @xiang90: have changed to use their copyright.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**